<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Manual &#8212; brie 0.1.3 documentation</title>
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="_static/bootstrap-3.3.7/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="_static/bootstrap-sphinx.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="FAQ" href="faq.html" />
    <link rel="prev" title="Installation" href="install.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="index.html">
          BRIE</a>
        <span class="navbar-text navbar-version pull-left"><b>0.1.3</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Manual</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="release.html">History</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Manual</a><ul>
<li><a class="reference internal" href="#brie-isoform-estimate">1. BRIE isoform estimate</a></li>
<li><a class="reference internal" href="#differential-splicing">2. Differential splicing</a></li>
<li><a class="reference internal" href="#splicing-events">3. Splicing events</a></li>
<li><a class="reference internal" href="#sequence-features">4. Sequence features</a></li>
<li><a class="reference internal" href="#preprocess">5. Preprocess</a><ul>
<li><a class="reference internal" href="#reads-alignment">5.1 reads alignment</a></li>
</ul>
</li>
<li><a class="reference internal" href="#examples">6. Examples</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
              
                
  <li>
    <a href="install.html" title="Previous Chapter: Installation"><span class="glyphicon glyphicon-chevron-left visible-sm"></span><span class="hidden-sm hidden-tablet">&laquo; Installation</span>
    </a>
  </li>
  <li>
    <a href="faq.html" title="Next Chapter: FAQ"><span class="glyphicon glyphicon-chevron-right visible-sm"></span><span class="hidden-sm hidden-tablet">FAQ &raquo;</span>
    </a>
  </li>
              
            
            
            
            
              <li class="hidden-sm">
<div id="sourcelink">
  <a href="_sources/manual.rst.txt"
     rel="nofollow">Source</a>
</div></li>
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12 content">
      
  <div class="section" id="manual">
<h1>Manual<a class="headerlink" href="#manual" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="table-of-contents">
<p class="topic-title first">Table of Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#brie-isoform-estimate" id="id1">1. BRIE isoform estimate</a></li>
<li><a class="reference internal" href="#differential-splicing" id="id2">2. Differential splicing</a></li>
<li><a class="reference internal" href="#splicing-events" id="id3">3. Splicing events</a></li>
<li><a class="reference internal" href="#sequence-features" id="id4">4. Sequence features</a></li>
<li><a class="reference internal" href="#preprocess" id="id5">5. Preprocess</a></li>
<li><a class="reference internal" href="#examples" id="id6">6. Examples</a></li>
</ul>
</div>
<p>After properly installed BRIE Python package, five excutable binary files could
be run from command line directly: <code class="docutils literal"><span class="pre">brie</span></code>, <code class="docutils literal"><span class="pre">brie-diff</span></code>, <code class="docutils literal"><span class="pre">brie-factor</span></code>,
<code class="docutils literal"><span class="pre">brie-event</span></code>, and <code class="docutils literal"><span class="pre">brie-event-filter</span></code>. Details about these utilities are
shown below.</p>
<div class="section" id="brie-isoform-estimate">
<h2><a class="toc-backref" href="#id1">1. BRIE isoform estimate</a><a class="headerlink" href="#brie-isoform-estimate" title="Permalink to this headline">¶</a></h2>
<p>This is the main program to quitify the fraction of exon inclusion level. In
order to automatically learn the informative prior, the predictive features are
required. There are two ways to get the annotation and sequence features:</p>
<ol class="arabic simple">
<li>use our processed annotation file and according sequence features, which you
can download from <a class="reference external" href="https://sourceforge.net/projects/brie-rna/files/annotation/">here</a>. Currently, we produced data for <a class="reference external" href="https://sourceforge.net/projects/brie-rna/files/annotation/human/gencode.v25/">human</a> and <a class="reference external" href="https://sourceforge.net/projects/brie-rna/files/annotation/mouse/gencode.vM12/">mouse</a>.
We suggest align RNA-seq reads to the according version of reference genome.</li>
<li>generate the annotation and fetch the sequence features with the help of
<a class="reference external" href="https://brie-rna.sourceforge.io/manual.html#splicing-events">brie-event</a> and <a class="reference external" href="https://brie-rna.sourceforge.io/manual.html#sequence-features">brie-factor</a> by yourself</li>
</ol>
<p>Then you could input the feature file obtained above, and run it like this:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">brie</span> <span class="o">-</span><span class="n">a</span> <span class="n">AS_events</span><span class="o">/</span><span class="n">SE</span><span class="o">.</span><span class="n">gold</span><span class="o">.</span><span class="n">gtf</span> <span class="o">-</span><span class="n">s</span> <span class="n">Cell1</span><span class="o">.</span><span class="n">sorted</span><span class="o">.</span><span class="n">bam</span> <span class="o">-</span><span class="n">f</span> <span class="n">mouse_features</span><span class="o">.</span><span class="n">csv</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">o</span> <span class="n">out_dir</span> <span class="o">-</span><span class="n">p</span> <span class="mi">15</span>
</pre></div>
</div>
<p>By default, you will have three output files in the out_dir: <code class="docutils literal"><span class="pre">fractions.tsv</span></code>,
<code class="docutils literal"><span class="pre">weights.tsv</span></code> and <code class="docutils literal"><span class="pre">samples.csv.gz</span></code>.</p>
<ul class="simple">
<li>In <code class="docutils literal"><span class="pre">fractions.tsv</span></code>, there are 8 columns:<ul>
<li>column 1: transcript id</li>
<li>column 2: gene id</li>
<li>column 3: transcript length</li>
<li>column 4: reads counts for whole events</li>
<li>column 5: FPKM for each isoform</li>
<li>column 6: fraction for each isoform, called Psi</li>
<li>column 7: lower bound of 95% confidence interval of isoform fraction</li>
<li>column 8: higher bound of 95% confidence interval of isoform fraction</li>
</ul>
</li>
<li>In <code class="docutils literal"><span class="pre">weights.tsv</span></code>, there are the weights for the Bayesian regression, with
<cite>#Feature+2</cite> lines, involving each features, interpret and sigma (a hyperparameter).
There are two columns each line, including the label and the value.</li>
<li>In <code class="docutils literal"><span class="pre">sample.csv.gz</span></code>, there are the <a class="reference external" href="https://en.wikipedia.org/wiki/Markov_chain_Monte_Carlo">MCMC</a> samples of posterior distribution of
Psi. These samples are used to detect the differential splicing.</li>
</ul>
<p>There are more parameters for setting (<code class="docutils literal"><span class="pre">brie</span> <span class="pre">-h</span></code> always give the version you
are using)</p>
<div class="highlight-html"><div class="highlight"><pre><span></span>Usage: brie [options]

Options:
  -h, --help            show this help message and exit
  -a ANNO_FILE, --anno_file=ANNO_FILE
                        Annotation file for genes and transcripts in GTF or
                        GFF3
  -s SAM_FILE, --sam_file=SAM_FILE
                        Sorted and indexed bam/sam files, use &#39;,&#39; for
                        replicates e.g., rep1.sorted.bam,sam1_rep2.sorted.bam
  -o OUT_FILE, --out_file=OUT_FILE
                        Prefix of the output files with full path
  -f FACTOR_FILE, --factor_file=FACTOR_FILE
                        Features in csv.gz file to predict isoform expression.

  Optional arguments:
    -p NPROC, --nproc=NPROC
                        Number of subprocesses [default: 4]
    -w WEIGHT_FILE, --weight_file=WEIGHT_FILE
                        File with weights, an output of Brie.
    -y FTYPE, --ftype=FTYPE
                        Type of function target: FPKM, Y, Psi [default: Y].
    --fLen=FRAG_LENG    Two arguments for fragment length: mean and standard
                        diveation, default: auto-detected
    --bias=BIAS_ARGS    Three argments for bias correction:
                        BIAS_MODE,REF_FILE,BIAS_FILE(s). BIAS_MODE: unif,
                        end5, end3, both. REF_FILE: the genome reference file
                        in fasta format. BIAS_FILE(s): bias files from dice-
                        bias, use &#39;---&#39; for time specific files, [default:
                        unif None None]
    --sigma=_SIGMA      Sigma in Bayesian regression: the Gaussian standard
                        deviation of residues [default: Auto].
    --lambda=_LAMBDA    Lambda in Bayesian regression: the coeffiecient of L2
                        constrain on weights [default: 0.1].
    --mcmc=MCMC_RUN     Four arguments for in MCMC iterations:
                        save_sample,max_run,min_run,gap_run. Required:
                        save_sample =&lt; 3/4*mim_run. [default: 500 5000 1000 50]
</pre></div>
</div>
<p><strong>Hyperparamers</strong></p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">sigma</span></code> is the square rooted variance of Gaussian noise in Bayesian
regression. By default, it will learn it automatically. Alternatively, you
could set it with your experience, for example, 3 might be a good option.</li>
<li><code class="docutils literal"><span class="pre">lambda</span></code> is the constrain on weights of Bayesian regression. 0.1 is good
option in ENCODE data.</li>
<li><code class="docutils literal"><span class="pre">weight_file</span></code> is fixed weights for Bayesian regression. Therefore, the
prior is predicted from the input weight file and its sequence features.</li>
</ul>
</div>
<div class="section" id="differential-splicing">
<h2><a class="toc-backref" href="#id2">2. Differential splicing</a><a class="headerlink" href="#differential-splicing" title="Permalink to this headline">¶</a></h2>
<p>This command allows to detect differential splicing between many cells
pair-wisely, including just two cells, by calculating Bayes factor. You could
run it as follows:</p>
<p>For two cells (<code class="docutils literal"><span class="pre">-p</span> <span class="pre">1</span> <span class="pre">--minBF</span> <span class="pre">0</span></code> gives all events in the same order. Speed:
10-20 second with 1 CPU)</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">brie</span><span class="o">-</span><span class="n">diff</span> <span class="o">-</span><span class="n">i</span> <span class="n">cell1</span><span class="o">/</span><span class="n">samples</span><span class="o">.</span><span class="n">csv</span><span class="o">.</span><span class="n">gz</span><span class="p">,</span><span class="n">cell2</span><span class="o">/</span><span class="n">samples</span><span class="o">.</span><span class="n">csv</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">o</span> <span class="n">c1_c2</span><span class="o">.</span><span class="n">diff</span><span class="o">.</span><span class="n">tsv</span> <span class="o">-</span><span class="n">p</span> <span class="mi">1</span> <span class="o">--</span><span class="n">minBF</span> <span class="mi">0</span>
</pre></div>
</div>
<p>For many cells (gives events with <code class="docutils literal"><span class="pre">BF&gt;10</span></code>. Speed: 100 cells in ~10min with 30
CPUs)</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>fileList=cell1/samples.csv.gz,cell2/samples.csv.gz,cell3/samples.csv.gz,cell4/samples.csv.gz

brie-diff -i $fileList -o c1_c4.diff.tsv
</pre></div>
</div>
<p>Then you will have an output file with 15 columns:</p>
<ul class="simple">
<li>column1-2: transcript id and gene id</li>
<li>column3-4: cell 1 and cell 2 names (the folder names)</li>
<li>column5-6: prior of exon inclusion fraction for cell 1 and cell 2</li>
<li>column7-8: posterior of exon inclusion fraction for cell 1 and cell 2</li>
<li>column9-12: counts for inclusion and exclusion for cell1, and then cell 2</li>
<li>column13-14: probability of prior and posterior diff&lt;0.05</li>
<li>column 15: Bayes factor</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Bayes factor is different from p value in hypothesis test. A good threshold
could be <code class="docutils literal"><span class="pre">Bayes</span> <span class="pre">factor</span> <span class="pre">&gt;</span> <span class="pre">10</span></code> as differential splicing event between two
cells.</p>
</div>
<p>There are more parameters for setting (<code class="docutils literal"><span class="pre">brie-diff</span> <span class="pre">-h</span></code> always give the version
you are using):</p>
<div class="highlight-html"><div class="highlight"><pre><span></span>Usage: brie-diff [options]

Options:
-h, --help            show this help message and exit
-i IN_FILES, --inFiles=IN_FILES
                      Input files of Brie samples for multiple cells, comma
                      separated for each cell, e.g., cell1,cell2,cell3
-o OUT_FILE, --outFile=OUT_FILE
                      Output file with full path

Optional arguments:
  -p NPROC, --nproc=NPROC
                      Number of subprocesses [default: 4]
  -n BOOTSTRAP, --bootstrap=BOOTSTRAP
                      Number of bootstrap [default: 1000]
  --minBF=MINBF       Minimum BF for saving out, e.g., 3 or 10. If it is 0,
                      save all events [default: 10]
</pre></div>
</div>
</div>
<div class="section" id="splicing-events">
<h2><a class="toc-backref" href="#id3">3. Splicing events</a><a class="headerlink" href="#splicing-events" title="Permalink to this headline">¶</a></h2>
<p><strong>Splicing events generating from full annotation</strong></p>
<p>This program is modified from Yarden Katz’s Python package <a class="reference external" href="https://github.com/yarden/rnaseqlib">rnaseqlib</a>, with
supporting different input annotation formats, e.g., gtf, gff3 and ucsc table.
For example, you could download a full annotation file for mouse from <a class="reference external" href="ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M12/gencode.vM12.annotation.gtf.gz">GENCODE</a>.
Then, you can generate the splicing event by the following command:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">brie</span><span class="o">-</span><span class="n">event</span> <span class="o">-</span><span class="n">a</span> <span class="n">gencode</span><span class="o">.</span><span class="n">vM12</span><span class="o">.</span><span class="n">annotation</span><span class="o">.</span><span class="n">gtf</span>
</pre></div>
</div>
<p>Then in the same folder of the annotation file, there will a new folder
<code class="docutils literal"><span class="pre">AS_events</span></code>, where the file for skipping-exon events, i.e., <code class="docutils literal"><span class="pre">SE.gff3</span></code>
is located.</p>
<p>There are more parameters for setting (<code class="docutils literal"><span class="pre">brie-event</span> <span class="pre">-h</span></code> always give
the version you are using):</p>
<div class="highlight-html"><div class="highlight"><pre><span></span>Usage: brie-event [options]

Options:
  -h, --help            show this help message and exit
  -a ANNO_FILE, --anno_file=ANNO_FILE
                        The annotation files used in making the annotation.
                        You could input multiple files; use comma&#39;,&#39; as
                        delimiter.
  --anno_type=ANNO_TYPE
                        The type of each annotation file. Use one for all
                        files or set for each file. Use comma &#39;,&#39; as
                        delimiter. You could choose &#39;ucsc&#39;, &#39;gtf&#39;, &#39;gff3&#39;.
                        [default: gtf]
  -o OUTPUT_DIR, --output_dir=OUTPUT_DIR
                        Output directory.
  --flanking-rule=FLANKING_RULE
                        Rule to use when defining exon trios. E.g.
                        &#39;commonshortest&#39; to use the most common and shortest
                        regions are flanking exons to an alternative trio.
                        [default: commonshortest]
  --multi-iso           If passed, generates multi-isoform annotations. Off by
                        default.
  --genome-label=GENOME_LABEL
                        If given, used as label for genome in output files.
  --sanitize            If passed, sanitize the annotation. Off by default.
</pre></div>
</div>
<p><strong>Splicing events quality check</strong></p>
<p>As the annotation file is not perfect, there may be false splicing events
generated from above command line. Therefore, we provide another function
<code class="docutils literal"><span class="pre">brie-event-filter</span></code> to only keep high-quality events, and use informative
ids. Based on above <code class="docutils literal"><span class="pre">SE.gff3</span></code>, we could select the gold-quality splicing
event by following command line. Note, the reference genome sequence is also
required, for example, mouse <a class="reference external" href="ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M12/GRCm38.p5.genome.fa.gz">genome</a> sequence here.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">brie</span><span class="o">-</span><span class="n">event</span><span class="o">-</span><span class="nb">filter</span> <span class="o">-</span><span class="n">a</span> <span class="n">AS_events</span><span class="o">/</span><span class="n">SE</span><span class="o">.</span><span class="n">gff3</span> <span class="o">-</span><span class="n">anno_ref</span> <span class="n">gencode</span><span class="o">.</span><span class="n">vM12</span><span class="o">.</span><span class="n">annotation</span><span class="o">.</span><span class="n">gtf</span> <span class="o">-</span><span class="n">r</span> <span class="n">GRCm38</span><span class="o">.</span><span class="n">p5</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">fa</span>
</pre></div>
</div>
<p>Then you will find an output file as <code class="docutils literal"><span class="pre">AS_events/SE.gold.gff3</span></code>, which only
contains splicing events passing the following constrains:</p>
<ul class="simple">
<li>located on autosome and input chromosome</li>
<li>not overlapped by any other AS-exon</li>
<li>surrounding introns are no shorter than a fixed length, e.g., 100bp</li>
<li>length of alternative exon regions, say, between 50 and 450bp</li>
<li>with a minimum distance, say 500bp, from TSS or TTS</li>
<li>surrounded by AG-GT, i.e., AG-AS.exon-GT</li>
</ul>
<p>There are more parameters for setting (<code class="docutils literal"><span class="pre">brie-event-filter</span> <span class="pre">-h</span></code> always give
the version you are using):</p>
<div class="highlight-html"><div class="highlight"><pre><span></span>Usage: brie-event-filter [options]

Options:
  -h, --help            show this help message and exit
  -a ANNO_FILE, --anno_file=ANNO_FILE
                        The annotation file of SE events in gff3 format from
                        rnaseqlib.
  --anno_ref=ANNO_REF   The reference annotation file in gtf format.
  -r REFERENCE, --reference=REFERENCE
                        The genome reference sequence file in fasta format.
  -o OUT_FILE, --out_file=OUT_FILE
                        The prefix of out files.
  --as_exon_min=AS_EXON_MIN
                        the minimum length for the alternative splicing exon.
  --as_exon_max=AS_EXON_MAX
                        the maximum length for the alternative splicing exon.
  --as_exon_tss=AS_EXON_TSS
                        the minimum length for the alternative exon to TSS.
  --as_exon_tts=AS_EXON_TTS
                        the minimum length for the alternative exon to TTS.
  --add_chrom=ADD_CHROM
                        the extra chromosomes besides autosome, e.g.,
                        chrX,chrY,chrM
</pre></div>
</div>
</div>
<div class="section" id="sequence-features">
<h2><a class="toc-backref" href="#id4">4. Sequence features</a><a class="headerlink" href="#sequence-features" title="Permalink to this headline">¶</a></h2>
<p>With the splicing annotation file, a set of short sequence feature can be
calculated by command line <code class="docutils literal"><span class="pre">brie-factor</span></code>. Besides the annotation file,
it also requires genome sequence file (the same as above), and a <a class="reference external" href="http://compgen.cshl.edu/phast/">phast</a>
conservation file in <a class="reference external" href="https://genome.ucsc.edu/goldenpath/help/bigWig.html">bigWig</a> format. For human and mouse, you could
download it directly from UCSC browser: <a class="reference external" href="http://hgdownload.cse.ucsc.edu/goldenPath/mm10/phastCons60way/">mm10.60way.phastCons.bw</a>
and <a class="reference external" href="http://hgdownload.cse.ucsc.edu/goldenPath/hg38/phastCons100way/">hg38.phastCons100way.bw</a>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">In order to fetch data from the bigWig file, we use a utility <code class="docutils literal"><span class="pre">bigWigSummary</span></code>
that is provided from UCSC. You could download the binary file for linux from
here: <a class="reference external" href="http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/bigWigSummary">http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/bigWigSummary</a></p>
</div>
<p>Besides, we require that <code class="docutils literal"><span class="pre">bigWigSummary</span></code> can be access in the PATH environment.
You can do it by the following command lines, and put into the <code class="docutils literal"><span class="pre">.profile</span></code> or
<code class="docutils literal"><span class="pre">.bashrc</span></code> file.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">chmod</span> <span class="o">+</span><span class="n">x</span> <span class="o">~/</span><span class="n">ucsc</span><span class="o">/</span><span class="n">bigWigSummary</span>
<span class="n">export</span> <span class="n">PATH</span><span class="o">=</span><span class="s2">&quot;~/ucsc:$PATH&quot;</span>
</pre></div>
</div>
<p>Then, you could get the sequence features by <code class="docutils literal"><span class="pre">brie-factor</span></code>, for example,</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">brie</span><span class="o">-</span><span class="n">factor</span> <span class="o">-</span><span class="n">a</span> <span class="n">AS_events</span><span class="o">/</span><span class="n">SE</span><span class="o">.</span><span class="n">gold</span><span class="o">.</span><span class="n">gtf</span> <span class="o">-</span><span class="n">r</span> <span class="n">GRCm38</span><span class="o">.</span><span class="n">p5</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">c</span> <span class="n">mm10</span><span class="o">.</span><span class="mi">60</span><span class="n">way</span><span class="o">.</span><span class="n">phastCons</span><span class="o">.</span><span class="n">bw</span> <span class="o">-</span><span class="n">o</span> <span class="n">mouse_features</span><span class="o">.</span><span class="n">csv</span> <span class="o">-</span><span class="n">p</span> <span class="mi">10</span>
</pre></div>
</div>
<p>Then you will have the features stored in a <code class="docutils literal"><span class="pre">mouse_features.csv.gz</span></code> file,
where #`factors` * #`gene_ids` features values are saved.</p>
<p>There are more parameters for setting (<code class="docutils literal"><span class="pre">brie-factor</span> <span class="pre">-h</span></code> always give the
version you are using):</p>
<div class="highlight-html"><div class="highlight"><pre><span></span>Usage: brie-factor [options]

Options:
  -h, --help            show this help message and exit
  -a ANNO_FILE, --anno_file=ANNO_FILE
                        Annotation file for genes and transcripts in GTF or
                        GFF3
  -r REF_SEQ, --ref_seq=REF_SEQ
                        Genome sequence reference in fasta file.
  -c PHAST_FILE, --phastCons=PHAST_FILE
                        PhastCons conservation scores in bigWig file.
  -o OUT_FILE, --out_file=OUT_FILE
                        Output in csv file, [default: brieFactor.cvs]

  Optional arguments:
    -p NPROC, --nproc=NPROC
                        Number of subprocesses [default: 4]
    --MSA5ss=MSA_5SS    Mutiple sequence alignment file for 5&#39;splice-site. It
                        is from -4 to 7. As default, MSA is based on input 5
                        splice sites.
    --MSA3ss=MSA_3SS    Mutiple sequence alignment file for 3&#39;splice-site. It
                        is from -16 to 4. As default, MSA is based on input 3
                        splice sites.
</pre></div>
</div>
</div>
<div class="section" id="preprocess">
<h2><a class="toc-backref" href="#id5">5. Preprocess</a><a class="headerlink" href="#preprocess" title="Permalink to this headline">¶</a></h2>
<div class="section" id="reads-alignment">
<h3>5.1 reads alignment<a class="headerlink" href="#reads-alignment" title="Permalink to this headline">¶</a></h3>
<p>Usually, the initial RNA-seq reads is in <a class="reference external" href="https://en.wikipedia.org/wiki/FASTQ_format">fastq</a> format, without information of
where it comes from the genome location. BRIE, similar as DICEseq and MISO, it
requires RNA-seq reads aligned to genome sequence. It should be in sam/bam
format, after sorting and indexing.</p>
<p>There are quite a fewer aligner that allows mapping reads to genome reference
with big gaps, mainly caused by splicing. For example, you could use <a class="reference external" href="https://code.google.com/p/rna-star/">STAR</a> and
<a class="reference external" href="https://ccb.jhu.edu/software/hisat/index.shtml">HISAT</a>, which usually return good alignment quality.</p>
<p>You could run it like this (based on HISAT 0.1.5), which including alignment,
sort and index:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>($hisatDir/hisat -x $hisatRef -1 $fq_dir/&quot;$file&quot;_1.fq.gz -2 $fq_dir/&quot;$file&quot;_2.fq.gz --no-unal | samtools view -bS -&gt; $out_dir/$file.bam) 2&gt; $out_dir/$file.err
samtools sort $out_dir/$file.bam $out_dir/$file.sorted
samtools index $out_dir/$file.sorted.bam
</pre></div>
</div>
</div>
</div>
<div class="section" id="examples">
<h2><a class="toc-backref" href="#id6">6. Examples</a><a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<p>There are some examples available here:
<a class="reference external" href="https://sourceforge.net/projects/brie-rna/files/examples/">https://sourceforge.net/projects/brie-rna/files/examples/</a></p>
<ul class="simple">
<li>Example to quantify splicing with provided annotation (bash code and data):
<a class="reference external" href="http://ufpr.dl.sourceforge.net/project/brie-rna/examples/brie_quantify/brie-examples.zip">brie-examples.zip</a></li>
<li>Example to quantify splicing with provided annotation (bash code):
<a class="reference external" href="https://github.com/huangyh09/brie/blob/master/example/brie_demo.sh">brie_demo.sh</a></li>
<li>Example to generate splicing events and fetch sequence factors (bash codes):
<a class="reference external" href="https://github.com/huangyh09/brie/blob/master/example/anno_maker.sh">anno_maker.sh</a></li>
</ul>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2018, Yuanhua Huang.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.3.<br/>
    </p>
  </div>
</footer>
  </body>
</html>