===========
Quick start
===========

.. note::
   This tutorial is mainly for alternative splicing analysis. If you are 
   interested in detecting **differential momentum genes (DMG)** for RNA 
   velocity analysis, please go to 
   `step 2.2 <quick_start.html#step2-2-phenotype-detection>`_ directly
   and also follow the tutorials on `scVelo <https://scvelo.org>`_.


BRIE estimates the isoform proportion for two-isoform events across many single 
cells. For getting started quickly, there are two main steps to go. A demo file
is available at `brie2_demo.sh`_. Sepcifically, you can follow the steps below.

.. _brie2_demo.sh: https://github.com/huangyh09/brie/blob/master/examples/brie2_demo.sh


Step1: read counts
==================

First, you need to count the isoform specific reads in each splicing events in 
each cell. For alternative splicing, e.g., exon-skipping event, you can download 
the splicing `annotations`_ generated by us or make your own, e.g., with 
`briekit`_.

Then you can use the ``brie-count`` fetch the read count tensor, which will be 
stored in hdf5 format as `AnnData`_. See more details on `brie-count 
CLI <brie_count.html>`_, and you can use this example command line.

.. code-block:: bash

  brie-count -a AS_events/SE.gold.gtf -S sam_and_cellID.tsv -o out_dir -p 15
  
.. _annotations: https://sourceforge.net/projects/brie-rna/files/annotation
.. _briekit: https://github.com/huangyh09/briekit/wiki
.. _AnnData: https://anndata.readthedocs.io

Besides the SE event, other types of alternative splicing, e.g., intron 
retaining is also applicable with BRIE. Some pre-processing utilities will be 
available soon.


Step2: quantification
=====================

`brie-quant` is used for this step. If GPU is available, we highly 
recommend using GPU for ~10x speedup comparing CPU server. Thanks to well 
support from Tensorflow, the environment setting is straightforwad, see our 
guide on `GPU Usage <./install.html#gpu-usage>`_.

Step2.1: isoform quantification
-------------------------------

Second, you can quantify the isoform with cell / gene or none features. Usually,
we recommend to use aggregated imputation even if you don't have any feature, 
namely ``mode 2`` in `brie-quant CLI <brie_quant.html>`_ as follows (please add 
``--interceptMode gene``),

.. code-block:: bash

  brie-quant -i out_dir/brie_count.h5ad -o out_dir/brie_quant_aggr.h5ad --interceptMode gene

Step2.2 phenotype detection
---------------------------

If you have cell level features, e.g., disease condition or cell type or 
continuous variable, you can use it in cell features to detect variable splicing
events or differential momentum genes as phenotypes for further analysis. This 
is ``mode 3`` in ``brie-quant``, so requires ``-c`` and ``--LRTindex``

.. code-block:: bash

  brie-quant -i out_dir/brie_count.h5ad -o out_dir/brie_quant_cell.h5ad \
      -c $DATA_DIR/cell_info.tsv --interceptMode gene --LRTindex=All


Step3: downstream analysis
==========================

The BRIE output AnnData files are compatible with `Scanpy`_, hence you can 
easily use it for dimension reduction, clustering, and other visualization. 
Example analyses are coming soon.

.. _Scanpy: https://scanpy.readthedocs.io

